### DISCUSSION TITLE
How does lazy load work on opt = true

### DISCUSSION THREAD

**Discussion by ekickx** [wbthomason/packer.nvim#425] [Q&A] [answered]
Hello,
I'm still a little confused by how lazy load works with `opt` set to `true`. Let say I have this config:

```lua
use {'nvim-telescope/telescope.nvim', keys = '<Leader>f', opt = true}
```

Does it mean that `packer` will do `packadd` when I press `<Leader>f` or I need to add `config = vim.cmd[[packadd telescope.nvim]]`? If packer will do `packadd` automatically, does it mean that:

```lua
use {'nvim-telescope/telescope.nvim', keys = '<Leader>f'}
-- and
use {'nvim-telescope/telescope.nvim', keys = '<Leader>f', opt = true}
```
will have the same behavior?

**Comment by gegoune** [comment, marked as answer]
Yes, keys implies opt, so it is the same.

**Reply by ekickx** [reply]
Ah ok. Thank you

**Reply by ekickx** [reply]
Ahh I understand it now. So the package that will be lazy loaded is installed in `opt` right.

**Reply by wbthomason** [reply]
Yep! You almost never need to manually specify `opt = true`. The main use cases for manually specifying that are (1) you want to track updates to a plugin but not use it, and (2) you want some custom lazy-loader logic that `packer` doesn't support.

**Reply by AckslD** [reply]
I guess a third case is if you have a plugin `B` that's only needed by another lazy loaded one `A`. So you say that
```
use {'A', cmd = 'Cmd', wants = 'B', requires = {{'B', opt = true}}
```

Or is there some other recommended way to do this?

**Reply by wbthomason** [reply]
@AckslD You raise an interesting point. We have optness-propagation for the `requires` key, but this should also apply to the `wants` key.

As it is, in your particular example, I would expect `B` to be opt even without the `opt = true` specifier because of this transitive optness. Is it not?

**Reply by AckslD** [reply]
@wbthomason I don't think it is actually, at least from what I've seen. See for example #477 where I wrote the table wrong so`opt` was not set and then the required plugin was installed in start.

**Reply by wbthomason** [reply]
Interesting. I think I would consider that a bug, but I'll have to look back at any discussion from when we added transitive optness to make sure this isn't intentional. Thanks!

**Reply by AckslD** [reply]
Cool, I'd be interested to hear what you find out :) Opt-ness propagation does seem natural to me 👍
